File: objectiveai-rs/src/agent/completions/http.rs
//! HTTP functions for agent completions.

use crate::{HttpClient, HttpError};
use futures::Stream;

/// Creates a agent completion (non-streaming).
///
/// Sends a request to the agent completions endpoint and waits for the
/// complete response.
///
/// # Arguments
///
/// * `client` - The HTTP client to use
/// * `params` - Agent completion parameters (stream flag will be set to false)
///
/// # Returns
///
/// The complete agent completion response.
pub async fn create_agent_completion_unary(
    client: &HttpClient,
    mut params: super::request::AgentCompletionCreateParams,
) -> Result<super::response::unary::AgentCompletion, HttpError> {
    params.stream = None;
    client
        .send_unary(reqwest::Method::POST, "agent/completions", Some(params))
        .await
}

/// Creates a streaming agent completion.
///
/// Sends a request to the agent completions endpoint and returns a stream
/// of response chunks as they arrive via Server-Sent Events.
///
/// # Arguments
///
/// * `client` - The HTTP client to use
/// * `params` - Agent completion parameters (stream flag will be set to true)
///
/// # Returns
///
/// A stream of agent completion chunks.
pub async fn create_agent_completion_streaming(
    client: &HttpClient,
    mut params: super::request::AgentCompletionCreateParams,
) -> Result<
    impl Stream<
        Item = Result<
            super::response::streaming::AgentCompletionChunk,
            HttpError,
        >,
    >
    + Send
    + 'static
    + use<>,
    HttpError,
> {
    params.stream = Some(true);
    client
        .send_streaming(reqwest::Method::POST, "agent/completions", Some(params))
        .await
}

File: objectiveai-py/objectiveai/agent/completions/http.py
"""HTTP functions for agent completions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai.agent.completions.request import AgentCompletionCreateParams
    from objectiveai.agent.completions.response.streaming import AgentCompletionChunk
    from objectiveai.agent.completions.response.unary import AgentCompletion
    from objectiveai.client import ObjectiveAI
    from objectiveai.stream import Stream


async def create_agent_completion(
    client: ObjectiveAI,
    params: AgentCompletionCreateParams,
) -> Union[AgentCompletion, Stream[AgentCompletionChunk]]:
    """Create an agent completion.

    If ``params.stream`` is true, returns a streaming response.
    Otherwise returns the complete response.
    """
    if getattr(params, "stream", None):
        return await client.post_streaming("agent/completions", params)
    return await client.post_unary("agent/completions", params)
